src/AMSlib/wf/basedb.hpp

```diff
@@ -1538,48 +1459,70 @@ class RabbitMQDB final : public BaseDB<TypeValue>
-    // "[err] buffer.c:1066: Assertion chain || datlen==0 failed in
-    // evbuffer_copyout" potentially segfault, and with CUDA could also lead to
-    // packet losses
-    std::this_thread::sleep_for(std::chrono::milliseconds(2000));
-
-    const size_t outputs_size = num_elements * outputs.size();
-    auto outputs_data = flatten_features(num_elements, outputs, true);
-    DBG(RabbitMQDB,
-        "[store(%d, %d, %d)] output sent %d B",
-        num_elements,
-        inputs.size(),
-        outputs.size(),
-        outputs_size * sizeof(TypeValue))
-    _nb_msg_send++;
-    _evbuffer->push(static_cast<void*>(outputs_data),
-                    outputs_size * sizeof(TypeValue));
-
-    free(inputs_data);
-    free(outputs_data);
+    const size_t num_in = inputs.size();
+    const size_t num_out = outputs.size();
+    const size_t header_size = 2*sizeof(uint8_t) + sizeof(uint32_t) + 3*sizeof(uint16_t);
+    const size_t blob_size = header_size + (num_elements * (num_in + num_out)) * sizeof(TypeValue);
+    uint8_t* data_blob = static_cast<uint8_t*>(std::malloc(blob_size*sizeof(uint8_t)));
+
+    uint16_t mpirank = static_cast<uint16_t>(_rank);
+    uint32_t num_elem = static_cast<uint32_t>(num_elements);
+    uint16_t input_dim = static_cast<uint16_t>(num_in);
+    uint16_t output_dim = static_cast<uint16_t>(num_out);
+
+    size_t current_offset = 0;
+    data_blob[current_offset] = static_cast<uint8_t>(header_size); // Header size (should be 12 bytes here)
+    current_offset += sizeof(uint8_t);
+
+    data_blob[current_offset] = static_cast<uint8_t>(sizeof(TypeValue)); // Data type (should be 1 bytes)
+    current_offset += sizeof(uint8_t);
+
+    std::memcpy(data_blob+current_offset, &(mpirank), sizeof(uint16_t)); // MPI rank (should be 2 bytes)
+    current_offset += sizeof(uint16_t);
+
+    std::memcpy(data_blob+current_offset, &(num_elem), sizeof(uint32_t)); // Num elem (should be 4 bytes)
+    current_offset += sizeof(uint32_t);
+
+    std::memcpy(data_blob+current_offset, &(input_dim), sizeof(uint16_t)); // Input dim (should be 2 bytes)
+    current_offset += sizeof(uint16_t);
+
+    std::memcpy(data_blob+current_offset, &(output_dim), sizeof(uint16_t)); // Output dim (should be 2 bytes)
+    current_offset += sizeof(uint16_t);
+
+    assert(current_offset == header_size);
+
+    for (size_t i = 0; i < num_elements; i++) {
+      for (size_t j = 0; j < num_in; j++) {
+        std::memcpy(data_blob+current_offset, &(inputs[j][i]), sizeof(TypeValue));
+        current_offset += sizeof(TypeValue); // We move on to the next TypeValue
+      }
+      for (size_t j = 0; j < num_out; j++) {
+        std::memcpy(data_blob+current_offset, &(outputs[j][i]), sizeof(TypeValue));
+        current_offset += sizeof(TypeValue);
+      }
+    }
+    assert(current_offset == blob_size);
+  
+    _evbuffer->push(data_blob, blob_size);
+    _pushed_msg++;
+    if (event_get_version_number() <= 0x02001500L) {
```

Comment:
can we add an assertion here? Also make it a hard requirement in CMAKE to require at least a version off...

Reply:
It would be best to put that in the CMake as this version of libevent is really old (2012) but that's the current version installed on Lassen TOSS3, that's why I just created an if, so we can still run the code.